docs/architecture/architecture.md

```diff
@@ -0,0 +1,683 @@
+not expected to work and may lead to inconsistencies as the block device will
+be simultaneously written to, from two different mounts. The copy operation from
+container to host will work, provided the user calls `sync(1)` from within the
+container prior to the copy to make sure any outstanding cached data is written
+to the block device.
+
+```
+docker cp [OPTIONS] CONTAINER:SRC_PATH HOST:DEST_PATH
+docker cp [OPTIONS] HOST:SRC_PATH CONTAINER:DEST_PATH
+```
+
+Clear Containers has the ability to hotplug block devices, which makes it
+possible to use block devices for containers started after the VM has been launched.
+
+Users can check to see if the container uses devicemapper block device as its
+rootfs by calling `mount(8)` within the counter.  If devicemapper block device
+is used, '/' will be mounted on `/dev/vda`.
+
+# Appendices
+
+## DAX
+
+Clear Containers utilises the Linux kernel DAX [(Direct Access filesystem)]
+(https://git.kernel.org/cgit/linux/kernel/git/torvalds/linux.git/tree/Documentation/filesystems/dax.txt)
+feature to efficiently map some host side files into the guest VM space.
+In particular, Clear Containers uses the `QEMU` nvdimm feature to provide a
+memory mapped virtual device that can be used to DAX map the virtual machine's
+root filesystem into the guest space.
+
+Mapping files using DAX provides a number of benefits over more traditional VM
+file and device mapping mechanisms:
+
+- Mapping as a direct access devices allows the guest to directly access
+the host memory pages (such as via eXicute In Place (XIP)), bypassing the guest
+page cache. This provides both time and space optimisations.
+- Mapping as a direct access device inside the VM allows pages from the
+host to be demand loaded using page faults, rather than having to make requests
+via a virtualised device (causing expensive VM exits/hypercalls), thus providing
+a speed optimisation.
+- Utilising MAP_SHARED shared memory on the host allows the host to efficiently
+share pages.
+
+Clear Containers uses the following steps to set up the DAX mappings:
+1. QEMU is configured with an nvdimm memory device, with a memory file
+  backend to map in the host side file into the virtual nvdimm space.
+2. The guest kernel command line mounts this nvdimm device with the DAX
+  feature enabled, allowing direct page mapping and access, thus bypassing the
+  guest page cache.
+
+![DAX](arch-images/DAX.png)
+
+Information on the use of nvdimm via QEMU is available in the [QEMU source code]
+(http://git.qemu-project.org/?p=qemu.git;a=blob;f=docs/nvdimm.txt;hb=HEAD)
+
+## Previous Releases
+
+This section provides a brief overview of architectural features enabled in
+prior and current versions of Clear Containers.
+
+### Version 3.0
+
+- Moved from hyperstart to `cc-agent` as an agent inside the VM.
+- Moved from `qemu-lite` to `pc` QEMU machine type.
+- Rewrite of runtime in go, leveraging virtcontainers.
```

Comment:
I'd add a few items here:

- virtio-blk for block based graph drivers
- New simplified protocol between the shim, proxy and runtime
- KSM throttling

Reply:
done.